### ISSUE TITLE
Tests excluded by grep get unsupported state in the JSON report

### ISSUE THREAD

**Issue by torgeilo** [plasticrake/mocha-json-serialize-reporter#21]
When using the grep option to run a subset of the tests (`?grep=foo` in the browser), all tests still end up in the serialized JSON. It makes mocha-json-runner unhappy because the skipped tests have no status and are not marked as pending:

```
Error: Unexpected test.state: undefined and not pending. test: MyTest
    at node_modules/mocha-json-runner/lib/mocha-json-runner.js:85:23
    at Array.forEach (<anonymous>)
    at runSuite (node_modules/mocha-json-runner/lib/mocha-json-runner.js:66:21)
    at node_modules/mocha-json-runner/lib/mocha-json-runner.js:97:11
    at Array.forEach (<anonymous>)
    at runSuite (node_modules/mocha-json-runner/lib/mocha-json-runner.js:96:22)
    at MochaJsonRunner.run (node_modules/mocha-json-runner/lib/mocha-json-runner.js:104:7)
```

`describe.skip`, `describe.only` etc work as expected (as the skipped tests are either marked as pending or not included in the result).

**Comment by plasticrake** [comment]
This is a much harder problem to solve that it should be. When I originally designed this I did not know there would be so many edge cases. Right now it inspects the test suite object after all the tests are run. However the test suite tree doesn't have enough detail to understand when things were not run via grep or when skipped due to failing hooks. I was able to make some good guesses about failing hooks, but grep is still hard. Right now I'm disabling the check that was throwing the error by default. 

Ideally I would look into building my own output stream based on the events emitted by the mocha runner instead of how I'm doing it now. But that would be almost an entire rewrite. Hopefully what was changed is a good enough effort. I'm also hampered somewhat in that I need this reporter to run on mocha v4 for my use case.
